misc/pragmas.md

```diff
@@ -0,0 +1,550 @@
+
+A conventional system for reading attribute grammars and making
+parsers which parse input and calculate the values of grammatical
+attributes might represent this grammar thus, naming the grammatical
+attribute *v* (this example follows, roughly the syntax of Alblas 191,
+and like Alblas assumes whitespace is someone else's problem).
+
+````
+    expr_0 →  expr_1 '+' term.
+            [ expr_0.v = expr_1.v + term.v]
+
+    term_0 →  term_1 '*' factor.
+            [ term_0.v = term_1.v * factor.v]
+
+    factor →  '0'.
+            [ factor.v = 0 ]
+	factor → '1'.
+	        [ factor.v = 1 ]
+	factor → '2'.
+            [ factor.v = 2 ]
+    factor → '3'.
+            [ factor.v = 3 ]
+    factor → '4'.
+            [ factor.v = 4 ]
+    factor → '5'.
+            [ factor.v = 5 ]
+    factor → '6'.
+            [ factor.v = 6 ]
+    factor → '7'.
+            [ factor.v = 7 ]
+    factor → '8'.
+            [ factor.v = 8 ]
+    factor → '9'.
+            [ factor.v = 9 ]
+    factor → '(' expr ')'.
+            [ factor.v = expr.v ]
+````
+
+Note that some nonterminals are subscripted so that references to
+their grammatical attributes can be unambiguous. To express this
+grammar in ixml, we need either to allow multiple rules for the same
+nonterminal, or to allow pragmas before connectors like comma or
+semicolon, or we need to allow string-to-typed-value functions in the
+style of XPath. I'll assume the latter two, along with a string()
+function that returns the string value of a nonterminal. With these
+assumptions, and the assumption that by means not specified the
+prefix *ag* has been bound to an appropriate grammar for
+attribute-grammar functionality, the attribute grammar could be
+written thus using the brackets-QName syntax
+
+````
+    [@ag:id e0] expr:  [@ag:id e1] expr, s, '+', s, term
+            [ag:rule e0.v := e1.v + term.v ].
+    {@ag:id t0] term:  [@ag:id t1] term, s, '_', s, factor
+            [ag:rule t0.v := t1.v * factor.v].
+	factor:  digit [@ag:rule factor.v := number(string(digit))]; 
+            '(', s, expr, s, ')' [@ag:rule factor.v := expr.v ].
+    digit: '0'; '1'; '2'; '3'; '4'; '5'; '6'; '7'; '8'; '9'.
+    s:  [#20; #A; #D; #9]_.
+````
+
+Here *ag:id* is assumed to associate a unique identifier with a
+particular instance of a nonterminal, and *ag:rule* is assumed to
+contain a set of assignment statements specifying the values of
+particular attributes, in a subset of XPath syntax. (A more serious
+proposal would need some way to distinguish *e0.v* meaning "the *v*
+attribute of *e0* from the same string as a name which happens to
+contain a string. This is not that serious proposal.)
+
+
+*Example:  synthesized value attribute and inherited environment attribute with variable bindings, for arithmetic expressions with 'let'.*
+
+
+## Open issues
+
+* Is it really OK not to have a way to have pragmas turn into elements
+before the first `rule` element?
```

Comment:
My understanding is that it's possible to have pragma turn into elements before the first `rule`, if they are defined before the first rule name.

What irks me is that this is inconsistent with the same use on other rules.

I would prefer something that is syntactically consistent.  How about allowing pragma inbetween rules, delimited by the period character.  The effect before a rule definition is for that pragma to apply as the immediately preceding node in the vxml tree?

I think it would mean that any such definitions which are ~serialised~ marked as attributes are collected together and applied to the grammar as a whole (represented as attributes on the root `ixml` element) to avoid attribute nodes being created after children nodes, but I could live with that.